src/clarity-angular/data/datagrid/providers/selection.ts

```diff
@@ -22,18 +22,53 @@ export enum SelectionType {
 @Injectable()
 export class Selection {
     public id: string;
+    private selected: any[] = [];  // Refs of selected items
+    private selectedSingle: any;   // Ref of single selected item
 
     constructor(private _items: Items, private _filters: FiltersProvider) {
         this.id = "clr-dg-selection" + (nbSelection++);
 
-        this._filtersSub = this._filters.change.subscribe(() => {
+        this.subscriptions.push(this._filters.change.subscribe(() => {
             if (!this._selectable) {
                 return;
             }
             this.clearSelection();
-        });
+        }));
 
-        this._itemsSub = this._items.allChanges.subscribe((updatedItems) => {
+        this.subscriptions.push(this._items.change.subscribe(updatedItems => {
+            if (!this._selectable || this._items.smart) {
+                return;
+            }
+            const leftOver: any[] = this.current.slice();
+            let newSingle: any;
+            if (this._items.trackBy) {
+                const trackBy: TrackByFunction<any> = this._items.trackBy;
+                updatedItems.forEach((item, index) => {
+                    const ref = trackBy(index, item);
+                    // Look in current selected refs array if item is selected, and update actual value
+                    if (this.selectedSingle === ref) {
+                        newSingle = item;
+                    } else if (this.selected.length) {
+                        const selectedIndex = this.selected.indexOf(ref);
+                        if (selectedIndex > -1) {
+                            leftOver[selectedIndex] = item;
+                        }
+                    }
+                });
+                // TODO: Discussed this with Eudes and this is fine for now.
+                // But we need to figure out a different pattern for the
+                // child triggering the parent change detection problem.
+                // Using setTimeout for now to fix this.
+                setTimeout(() => {
+                    if (typeof newSingle !== "undefined") {
+                        this.currentSingle = newSingle;
+                    }
+                    this.current = leftOver;
+                }, 0);
+            }
+        }));
+
+        this.subscriptions.push(this._items.allChanges.subscribe((updatedItems) => {
             if (!this._selectable) {
```

Comment:
The two subscriptions look very similar, I'm wondering if we can avoid code duplication.

And saying this just made me realize there is a final bug left here: use a client-side datagrid with a "heartbeat", that refreshes the complete set of items every X seconds.
In this specific case, we're still using the wrong index in the trackBy of the `this._items.allChanges` subscription.

I believe both remarks can be solved by putting the new code you wrote inside the `this._items.allChanges` subscription, and update the Items provider to:
```typescript
private _all: any[];
public set all(items: any[]) {
    if (this.smart) {
        this._all = items;
        this.emitAllChanges();
        this._filterItems();
    } else {
        this._displayed = items;
        this.emitAllChanges();
        this.emitChange();
    }
}
```